Check other rvalues

spastorino opened the pull request:
#15 but done on top of `nll-master`

This should close https://github.com/rust-lang/rust/issues/45827 and https://github.com/rust-lang/rust/issues/45959.

nikomatsakis committed:
fix filter to support `&` and `|`

spastorino committed:
Check NullaryOp Rvalue

spastorino committed:
Check Aggregate predicates

nikomatsakis committed:
fix universal regions to handle constant expressions like `[T; 22]`

spastorino committed:
Check functions predicates

spastorino committed:
rustfmt: MIR typeck

spastorino committed:
Check Repeat Rvalue

spastorino committed:
Add more debug logs

nikomatsakis commented on src/librustc_mir/transform/type_check.rs:
```diff
@@ -59,7 +59,7 @@ pub fn type_check<'a, 'gcx, 'tcx>(
 
 fn mirbug(tcx: TyCtxt, span: Span, msg: &str) {
-    tcx.sess.diagnostic().span_bug(span, msg);
+    tcx.sess.diagnostic().delay_span_bug(span, msg);
```
maybe worth a comment here -- something like:

```
// We sometimes see MIR failures (notably predicate failures) due to
// the fact that we check rvalue sized predicates here. So use `delay_span_bug`
// to avoid reporting bugs in those cases.
```

nikomatsakis commented on src/librustc_mir/transform/type_check.rs:
```diff
@@ -1240,6 +1242,31 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
+
+            AggregateKind::Closure(def_id, substs) |
+            AggregateKind::Generator(def_id, substs, _) => {
+                tcx.predicates_of(*def_id).instantiate(tcx, substs.substs)
```
I realized that [this code](https://github.com/spastorino/rust/blob/017b3c5fe90b846da610b9372977517f8b65a275/src/librustc_mir/transform/type_check.rs#L1179-L1207) should probably move into this match arm

nikomatsakis reviewed (changes requested):
Lookin' good. Two nits.

spastorino committed:
Restructure a bit check_aggregate_rvalue code

spastorino commented:
@nikomatsakis addressed both things

nikomatsakis commented on src/librustc_mir/transform/type_check.rs:
```diff
@@ -1210,45 +1210,13 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
-        self.prove_aggregate_predicates(aggregate_kind, location);
-
-        match aggregate_kind {
+        if *aggregate_kind == AggregateKind::Tuple {
```
Nit: if we are going to "return early" for tuples, we should have `bug!` cases for the `prove_aggregate_predicates` function etc. Otherwise if people tried to add code there there might be confusion.

spastorino replied:
The other thing we could do is to leave prove_aggregate_predicates before the early return as it was ... so if you want to handle Tuples there you can.

nikomatsakis replied:
Either would be fine.

spastorino replied:
Done!

spastorino committed:
Mir typeck Cast for ReifyFnPtr value

spastorino committed:
Mir typeck Cast for UnsafeFnPtr value

spastorino committed:
Extract coerce_closure_fn_ty function

spastorino committed:
Mir typeck Cast for ClosureFnPtr value

spastorino committed:
Mir typeck Cast for Unsize value

nikomatsakis committed:
normalize fn sig as part of reification

nikomatsakis committed:
when reifying a safe fn as an unsafe fn ptr, insert two casts

Otherwise, `run-pass/typeck-fn-to-unsafe-fn-ptr.rs` fails the MIR type checker.

nikomatsakis committed:
fix closure tests now that MIR typeck works properly

These tests had FIXMEs for errors that were not previously being
reported.
